src/main/cpp/src/map_utils.cu

```diff
@@ -0,0 +1,633 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "map_utils_debug.cuh"
+
+//
+#include <limits>
+
+//
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/null_mask.hpp>
+#include <cudf/detail/utilities/algorithm.cuh>
+#include <cudf/detail/utilities/vector_factories.hpp>
+#include <cudf/io/detail/tokenize_json.hpp>
+#include <cudf/strings/detail/combine.hpp>
+#include <cudf/strings/detail/strings_children.cuh>
+#include <cudf/strings/string_view.hpp>
+#include <cudf/strings/strings_column_view.hpp>
+
+//
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+
+//
+#include <thrust/binary_search.h>
+#include <thrust/copy.h>
+#include <thrust/count.h>
+#include <thrust/find.h>
+#include <thrust/for_each.h>
+#include <thrust/functional.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/iterator/permutation_iterator.h>
+#include <thrust/iterator/transform_iterator.h>
+#include <thrust/pair.h>
+#include <thrust/scan.h>
+#include <thrust/sequence.h>
+#include <thrust/transform.h>
+#include <thrust/transform_reduce.h>
+
+//
+#include <cub/device/device_radix_sort.cuh>
+
+namespace spark_rapids_jni {
+
+using namespace cudf::io::json;
+
+namespace {
+
+// Unify the input json strings by:
+// 1. Append one comma character (',') to the end of each input string, except the last one.
+// 2. Concatenate all input strings into one string.
+// 3. Add a pair of bracket characters ('[' and ']') to the beginning and the end of the output.
+rmm::device_uvector<char> unify_json_strings(cudf::column_view const &input,
+                                             rmm::cuda_stream_view stream) {
+  if (input.is_empty()) {
+    return cudf::detail::make_device_uvector_async<char>(std::vector<char>{'[', ']'}, stream);
+  }
+
+  auto const d_strings = cudf::column_device_view::create(input, stream);
+  auto const transform_it = thrust::make_counting_iterator(0);
+  auto const output_size =
```

Comment:
Why not look at the last offset for the total characters in the column? Is it due to nulls needing 2 characters? Is this faster than `null_count() * 2`?

Reply:
Hah, good point. I'll replace this with a trivial computation.